### PR TITLE
feat(setup): exclude runtime data from Spotlight

### PR DESCRIPTION
**Pull request by EtanHey**
## Summary

- create `.metadata_never_index`-backed roots for every high-churn BrainLayer data/runtime location before macOS setup or init starts services, including both canonical and overridden DB roots
- refuse to mutate nonempty legacy trees and make `brainlayer doctor` warn when the active data directory is not excluded or cannot be checked
- document a READY checkpoint/stop/migrate/restart/verify/rollback runbook with exact partial-state recovery, a complete writer fence, and a fail-closed same-volume positive control

## Safety

- the canonical live BrainLayer data directory was not migrated or modified
- the live migration is explicitly deferred to an approved writer-stop window
- release PRs #705 and #707 merged first; this final head incorporates the v1.5.6 mainline
- pair review V2 declared the shipped code mergeable but found runbook-only execution defects; the live runbook remains blocked pending the immediately-following docs-only V3 delta

## Verification

- focused Spotlight suites: 162 passed
- final exact-head pre-push gate: 3,986 passed, 10 skipped, 61 deselected, 2 xfailed
- MCP registration: 3 passed; isolated eval/hook routing: 40 passed; Bun: 1 passed; FTS5 shell regression: passed
- Ruff, format check, `git diff --check`: clean
- prior hosted code findings were addressed; ambiguous duplicate DB assignments and op-backed DB paths now fail preflight rather than choosing between launchd and direct-Python semantics

— brainlayer-worker-1ciwjp (worker) · codex/gpt-5.6-sol

---

> [!NOTE]
> **Medium Risk**
> Touches setup, launchd install, env DB loading, and path resolution on macOS; misconfiguration or legacy-tree refusal could block installs until migration, but live data is not modified by this PR.
> 
> **Overview**
> Adds **macOS Spotlight exclusion** for BrainLayer’s high-churn runtime trees using `.metadata_never_index`, wired into setup, launchd install, and doctor—without migrating live production data in this PR.
> 
> **Runtime behavior:** New `spotlight.py` layout preflight creates marker-backed roots (data, `~/.brainlayer`, launchd logs, P0 counter) before children, validates `BRAINLAYER_DB` override safety, and **refuses** to touch nonempty legacy trees (points operators at the migration runbook). `brainlayer setup` / `init` run this first on Darwin; `install.sh` runs the same Python preflight before `mkdir` for install actions (skipped for `remove`/`unload`/`load`). `paths` gains `is_spotlight_excluded`; `resolve_db_path` / `DEFAULT_DB_PATH` no longer create dirs at import.
> 
> **Doctor & config:** Darwin doctor warns on `spotlight_indexing_enabled` when the active DB directory isn’t excluded, and fatals on `brainlayer_db_config_invalid` when env `BRAINLAYER_DB` is ambiguous or launchd-incompatible—`load_brainlayer_env` now aligns DB parsing with launchd grammar and records fallback errors.
> 
> **Ops docs:** Adds design/plan docs and a **READY** writer-stop migration runbook (WAL checkpoint, full writer fence, stage/rename/mark/restore, fail-closed Spotlight probe with same-volume positive control, rollback).
> 
> **Tests:** Extensive pytest coverage for layout edge cases, CLI ordering, launchd preflight, config/DB validation, and runbook invariants.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit f8cb0b6f4ff697faf942a7dfd003ae870acdd26d. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

> [!NOTE]
> ### Exclude runtime data directories from macOS Spotlight indexing during setup
> - Adds `brainlayer.spotlight.ensure_spotlight_excluded_layout` to create and validate Spotlight-excluded root directories by placing `.metadata_never_index` markers; the function validates each root is disjoint, non-overlapping, and not under a symlink or shared parent.
> - Integrates the preflight into the `init` and `setup` CLI commands on Darwin, and into `scripts/launchd/install.sh` before any directory creation.
> - Adds `brainlayer.paths.is_spotlight_excluded` to check whether a path is under an excluded root, and `brainlayer.doctor` now warns when the data directory is not excluded and reports a fatal error when `BRAINLAYER_DB` config is ambiguous between launchd and direct parsing.
> - Adds `config.configured_brainlayer_env_value` to strictly parse and validate launchd-compatible env assignments, rejecting command substitutions, duplicates, and ambiguous quoting.
> - Risk: installs on macOS with pre-existing non-excluded runtime trees will fail the preflight and require a manual migration runbook before installation can proceed.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized f8cb0b6.</sup>
> 
>

### CODE REVIEW

File: scripts/launchd/install.sh [hunk 1]
```diff
@@ -82,6 +82,29 @@ if [ ! -x "$BRAINLAYER_BIN" ]; then
...
+                BRAINLAYER_PREFLIGHT_PYTHONPATH="$BRAINLAYER_DIR/src"
+            else
+                BRAINLAYER_PREFLIGHT_PYTHONPATH="$BRAINLAYER_DIR"
+            fi
+            if [ -x "$PYTHON_BIN" ]; then
+                BRAINLAYER_PREFLIGHT_PYTHON="$PYTHON_BIN"
+            else
+                BRAINLAYER_PREFLIGHT_PYTHON="$(command -v python3)"
+            fi
+            PYTHONPATH="$BRAINLAYER_PREFLIGHT_PYTHONPATH${PYTHONPATH:+:$PYTHONPATH}" \
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/launchd/install.sh:114]: Fixed in `ff38da63`. The standalone installer now resolves `BRAINLAYER_DB` from the selected `BRAINLAYER_ENV_FILE`, while preserving original process-environment precedence, before running the preflight. A real installer regression test proves the selected override root receives the marker.

— brainlayer-worker-1ciwjp (worker) · codex/gpt-5.6-sol

File: docs/operations/spotlight-exclusion-migration.md [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+First establish that Spotlight is enabled for the containing volume of every migration root. For
+each canonical root and the validated distinct override root, resolve the filesystem mount point,
+record the root-to-volume mapping, and query that mount point directly:
+
+```bash
+brainlayer_probe_root="${HOME:?}/.local/share/brainlayer"
+brainlayer_probe_volume="$(stat -f %m "$brainlayer_probe_root")"
+printf 'root=%s\nvolume=%s\n' "$brainlayer_probe_root" "$brainlayer_probe_volume"
+mdutil -s "$brainlayer_probe_volume"
+```
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply docs/operations/spotlight-exclusion-migration.md:257]: Fixed in `ff38da63`, with the current-macOS key corrected from the suggestion. On macOS 26.5.1, `stat -f %m` returned mtime, `diskutil info -plist <directory>` returned no disk, and the plist exposes `MountPoint` rather than `VolumeMountPoint`. The runbook now resolves the device with quoted `df -P`, extracts `MountPoint` from `diskutil info -plist <device>`, prints root/device/volume, and passes the quoted mount point to `mdutil -s`.

— brainlayer-worker-1ciwjp (worker) · codex/gpt-5.6-sol

File: src/brainlayer/spotlight.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    *,
+    home: Path,
+    ismount: Callable[[Path], bool],
+) -> None:
+    active = active_root.expanduser().resolve(strict=False)
+    canonical = canonical_root.expanduser().resolve(strict=False)
+    if active == canonical:
+        return
+    unsafe_location = active == home.resolve() or active == Path(active.anchor) or ismount(active)
+    if unsafe_location or active in canonical.parents or "brainlayer" not in active.name.casefold():
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/spotlight.py]: Fixed in `ff38da63`. Validation now walks every unresolved component of the expanded active override parent and rejects any symbolic link before canonical resolution. The regression uses a database below a symlinked ancestor and proves no override tree is created.

— brainlayer-worker-1ciwjp (worker) · codex/gpt-5.6-sol

File: tests/test_installable_build.py [hunk 4]
```diff
@@ -747,6 +1038,41 @@ def test_launchd_installer_preflights_all_before_loading_without_google_key(tmp_
...
+        capture_output=True,
+        text=True,
+        timeout=30,
+        check=False,
+    )
+
+    assert result.returncode == 73
+    assert "ensure_spotlight_excluded_layout" in preflight_log.read_text(encoding="utf-8")
+    assert not (home / "Library" / "LaunchAgents").exists()
+    assert not (home / ".local" / "share" / "brainlayer").exists()
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply tests/test_installable_build.py:1249]: Fixed in `ff38da63`. The test now prepends a `uname` shim that reports Darwin, so the macOS-only preflight executes on non-macOS CI while retaining the exit-code, call-log, and no-directory-creation assertions.

— brainlayer-worker-1ciwjp (worker) · codex/gpt-5.6-sol

File: src/brainlayer/config.py [hunk 5]
```diff
@@ -91,6 +92,27 @@ def _resolve_op_read_value(value_text: str) -> str | None:
     return result.stdout.rstrip("\n")
 
 
+def configured_brainlayer_env_value(name: str, env_path: Path | None = None) -> str | None:
+    """Return one configured value with original process-env precedence."""
+    if name in _PROCESS_ENV_AT_IMPORT:
+        return _PROCESS_ENV_AT_IMPORT[name]
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/config.py]: Fixed in `e3659579`. Preflight now intentionally reads the selected env file’s effective DB value instead of a transient inherited shell value, matching what `brainlayer-env-run.sh` loads for installed jobs. The regression passes a conflicting process override and proves only the selected env-file root is marked.

— brainlayer-worker-1ciwjp (worker) · codex/gpt-5.6-sol

File: src/brainlayer/config.py [hunk 6]
```diff
@@ -91,6 +92,27 @@ def _resolve_op_read_value(value_text: str) -> str | None:
...
+    except (OSError, UnicodeDecodeError):
+        return None
+
+    for line in lines:
+        assignment = _split_env_assignment(line)
+        if assignment is None or assignment[0] != name:
+            continue
+        value = _parse_env_value(assignment[1])
+        if value is not None:
+            return value
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/config.py]: Fixed in `e3659579`. The selected env-file lookup scans the complete file, skips invalid values, and retains the final valid assignment, matching the runtime env runner’s last-write behavior. The test covers invalid-first plus two valid assignments.

— brainlayer-worker-1ciwjp (worker) · codex/gpt-5.6-sol

File: scripts/launchd/install.sh [hunk 7]
```diff
@@ -82,6 +82,29 @@ if [ ! -x "$BRAINLAYER_BIN" ]; then
     exit 1
 fi
 
+BRAINLAYER_INSTALL_ACTION="${1:-all}"
+case "$BRAINLAYER_INSTALL_ACTION" in
+    remove|unload)
+        ;;
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply scripts/launchd/install.sh]: Fixed in `e3659579`. Runtime-directory creation now occurs only in install/load actions after preflight; `remove` and `unload` bypass both preflight and all mkdir calls. Parameterized fresh-home regressions prove neither teardown action creates data, runtime, or launchd-log roots.

— brainlayer-worker-1ciwjp (worker) · codex/gpt-5.6-sol

File: src/brainlayer/config.py [hunk 8]
```diff
@@ -91,6 +91,28 @@ def _resolve_op_read_value(value_text: str) -> str | None:
...
+        return None
+
+    selected: str | None = None
+    for line in lines:
+        assignment = _split_env_assignment(line)
+        if assignment is None or assignment[0] != name:
+            continue
+        if "$(" in assignment[1] or "`" in assignment[1]:
+            continue
+        value = _parse_env_value(assignment[1])
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/config.py]: Fixed at exact head 25351a79. Spotlight preflight now uses the launchd runner's simple assignment grammar: escaped spaces remain literal, one matching outer quote pair is removed, and malformed double-space export syntax is ignored. Regression coverage pins each case.

— brainlayer-worker-1ciwjp (worker) · codex/gpt-5.6-sol

File: src/brainlayer/config.py [hunk 9]
```diff
@@ -91,6 +91,25 @@ def _resolve_op_read_value(value_text: str) -> str | None:
...
+        return None
+
+    selected: str | None = None
+    for line in lines:
+        assignment = _split_env_assignment(line)
+        if assignment is None or assignment[0] != name:
+            continue
+        value = _parse_env_value(assignment[1])
+        if value is not None:
+            selected = value
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/config.py]: Strengthened at exact head 25351a79. A DB assignment that Python could resolve through op but launchd would skip is now an explicit preflight error, so setup cannot silently mark a different root. Unsupported substitutions remain ignored only when neither runtime can resolve them.

— brainlayer-worker-1ciwjp (worker) · codex/gpt-5.6-sol

[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/config.py]: Resolved fail-closed at exact head 25351a79. Duplicate valid DB assignments are now rejected because launchd is last-wins while direct Python loading is first-wins; preflight no longer guesses either root. Single assignments use the launchd runner's exact simple-value grammar.

— brainlayer-worker-1ciwjp (worker) · codex/gpt-5.6-sol

File: src/brainlayer/config.py [hunk 10]
```diff
@@ -91,6 +114,31 @@ def _resolve_op_read_value(value_text: str) -> str | None:
...
+    selected: str | None = None
+    for line in lines:
+        assignment = _split_launchd_env_assignment(line)
+        if assignment is None or assignment[0] != name:
+            continue
+        if _OP_READ_PREFIX in assignment[1]:
+            raise RuntimeError(f"{name} must not use an op command substitution in {target}")
+        if "$(" in assignment[1] or "`" in assignment[1]:
+            continue
+        value = _parse_launchd_env_value(assignment[1])
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/config.py:138]: Fixed in `d7235552`. The selected-env preflight now compares the launchd-literal value with the direct CLI parser result and fails closed when they differ. Regression coverage includes both escaped-space and nested-quote forms; quoted-space values that resolve identically remain accepted. The focused path/install surface passes 95 tests and the full pre-push gate passes at the exact head.

— brainlayer-worker-1ciwjp (worker) · codex/gpt-5.6-sol

File: src/brainlayer/spotlight.py [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+    home = home_fn()
+    if data_dir is not None:
+        data_roots = (data_dir,)
+    else:
+        canonical_data_root = get_canonical_db_path_fn().parent
+        if env_file is not None:
+            from .config import configured_brainlayer_env_value
+
+            configured_db = configured_brainlayer_env_value("BRAINLAYER_DB", env_file)
+            active_data_root = Path(configured_db).expanduser().parent if configured_db else canonical_data_root
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/spotlight.py]: Fixed in `c87ba7cc`. `resolve_db_path()` now expands `~` before applying the runtime guard, matching the Spotlight preflight. The regression test pins the same resolved path for runtime and preflight use.

— brainlayer-worker-1ciwjp (worker) · codex/gpt-5.6-sol

[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/spotlight.py]: Fixed in `c87ba7cc`. The preflight now rejects a configured DB file that is itself a symlink before deriving or creating any root. Existing ancestor-symlink rejection remains in place, and regression coverage proves no canonical root is created on refusal.

— brainlayer-worker-1ciwjp (worker) · codex/gpt-5.6-sol

File: src/brainlayer/spotlight.py [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+            active_data_root = resolve_db_path_fn().parent
+        _validate_override_data_root(active_data_root, canonical_data_root, home=home, ismount=ismount_fn)
+        data_roots = tuple(
+            dict.fromkeys(root.expanduser().resolve(strict=False) for root in (active_data_root, canonical_data_root))
+        )
+    requested_roots = (
+        *data_roots,
+        runtime_dir or home / ".brainlayer",
+        launchd_log_dir or home / "Library" / "Logs" / "brainlayer",
+        counter_dir or home / ".brainlayer-p0-counter",
```
[the remark this answers is not in the document]
Comment (EtanHey) [reply src/brainlayer/spotlight.py:137]: Fixed in `c87ba7cc`. Selected-env high-churn runtime path overrides, including `BRAINLAYER_QUEUE_DIR`, now fail closed before any marker/root creation. The guard list covers the durable queue, watcher state/quarantine, drain state/logs, enrichment offsets/costs, backup state/staging/logs, counter, and writer telemetry paths.

— brainlayer-worker-1ciwjp (worker) · codex/gpt-5.6-sol
